### PR TITLE
WIP: add preemptBasedOnFlavorOrder

### PR DESCRIPTION
**Pull request by KunWuLuan**
to #582

### CODE REVIEW

**Comment by KunWuLuan** [comment]
I will add test in another commit later.

**Comment by KunWuLuan** [comment]
@alculquicondor Hi, aldo. I just joined the Kubernetes slack. My username is KunWuLuan

File: apis/kueue/v1beta1/clusterqueue_types.go [hunk 1]
```diff
@@ -272,6 +272,9 @@ type ClusterQueuePreemption struct {
 	// +kubebuilder:default=Never
 	// +kubebuilder:validation:Enum=Never;LowerPriority
 	WithinClusterQueue PreemptionPolicy `json:"withinClusterQueue,omitempty"`
+
+	// +kubebuilder:default="flase"
+	PreemptBasedOnFlavorOrder bool `json:"preemptBasedOnFlavorOrder"`
```
Comment (alculquicondor) [inline apis/kueue/v1beta1/clusterqueue_types.go:277]: In general, booleans might not be the best API https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types

For example:

```yaml
  flavorFungibility: NextFlavorBeforePreemption | PreemptionBeforeNexFlavor
```

I don't really like the names I'm proposing, but I'll leave it to you to come up with other ideas.

File: pkg/scheduler/flavorassigner/flavorassigner.go [hunk 2]
```diff
@@ -368,6 +368,9 @@ func (a *Assignment) findFlavorForResourceGroup(
 			}
 		}
 
+		if cq.Preemption.PreemptBasedOnFlavorOrder && representativeMode == Preempt {
```
Comment (alculquicondor) [inline pkg/scheduler/flavorassigner/flavorassigner.go:371]: The idea should be that if preemption is not possible, we proceed to the next flavor. But the flavor assigner doesn't really know about whether preemption is possible.
We might need to change the implementation of framework assigner to return a list of assignments that the scheduler and preemption algorithm. Or maybe we need to add tighter dependencies between the flavorAssigner and the preemption algorithms.

**Review by alculquicondor — COMMENTED** [review]
/retitle WIP: add preemptBasedOnFlavorOrder

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go:371]: I think we can just keep track in-memory of whether preemption was possible. If it was not, we put the workload back into the head of the queue to try the next flavor in the next iteration. And once we have passed through all the flavors, we requeue the workload as inadmissible.

File: apis/kueue/v1beta1/clusterqueue_types.go (hunk 1 above)
Comment (KunWuLuan) [reply apis/kueue/v1beta1/clusterqueue_types.go:277]: how about 
```
PreemptionTiming: BeforeNextFlavor | AfterAllFlavors
```

File: pkg/scheduler/flavorassigner/flavorassigner.go (hunk 2 above)
Comment (KunWuLuan) [reply pkg/scheduler/flavorassigner/flavorassigner.go:371]: Thanks for your time and help. I will continue to improve this submission. 😸

Comment (alculquicondor) [reply pkg/scheduler/flavorassigner/flavorassigner.go:371]: consider writing a design for this feature, following this template https://github.com/kubernetes-sigs/kueue/tree/main/keps/NNNN-template

Comment (KunWuLuan) [reply pkg/scheduler/flavorassigner/flavorassigner.go:371]: No problem, I will submit a proposal in next week

**Comment by KunWuLuan** [comment]
@alculquicondor @mortent  Hi, I have submited a kep for this feature. If you have time, could you take a look and give me your opinion? Thank you!

**Comment by KunWuLuan** [comment]
@alculquicondor @mortent  Hi, I have submited a kep for this feature. If you have time, could you take a look and give me your opinions? Thank you!

**Comment by tenzen-y** [comment]
> @alculquicondor @mortent Hi, I have submited a kep for this feature. If you have time, could you take a look and give me your opinions? Thank you!

@KunWuLuan Thank you for submitting a KEP.
Can you create a separate PR for a KEP instead of including this PR?

**Comment by KunWuLuan** [comment]
> > @alculquicondor @mortent Hi, I have submited a kep for this feature. If you have time, could you take a look and give me your opinions? Thank you!
> 
> @KunWuLuan Thank you for submitting a KEP. Can you create a separate PR for a KEP instead of including this PR?

@tenzen-y OK, I will create a separate RP to track the KEP, thanks

**Comment by alculquicondor** [comment]
@KunWuLuan I think the other approach is better. Can we close this?

**Comment by KunWuLuan** [comment]
the other approach is better #849

**Pull request closed without merging**